Library/OcConsoleLib/GopPassThrough.c

```diff
@@ -0,0 +1,285 @@
+  UINTN                            HandleCount;
+  EFI_HANDLE                       *HandleBuffer;
+  UINTN                            Index;
+  EFI_GRAPHICS_OUTPUT_PROTOCOL     *GraphicsOutput;
+  EFI_UGA_DRAW_PROTOCOL            *UgaDraw;
+  OC_GOP_PROTOCOL                  *OcGopDraw;
+  APPLE_FRAMEBUFFER_INFO_PROTOCOL  *FramebufferInfo;
+  EFI_PHYSICAL_ADDRESS             FramebufferBase;
+  UINT32                           FramebufferSize;
+  UINT32                           ScreenRowBytes;
+  UINT32                           ScreenWidth;
+  UINT32                           ScreenHeight;
+  UINT32                           ScreenDepth;
+  UINT32                           HorizontalResolution;
+  UINT32                           VerticalResolution;
+  EFI_GRAPHICS_PIXEL_FORMAT        PixelFormat;
+  UINT32                           ColorDepth;
+  UINT32                           RefreshRate;
+
+  Status = gBS->LocateHandleBuffer (
+    ByProtocol,
+    &gEfiGraphicsOutputProtocolGuid,
+    NULL,
+    &HandleCount,
+    &HandleBuffer
+    );
+
+  if (!EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCC: Found %u handles with GOP draw\n", (UINT32) HandleCount));
+    FreePool (HandleBuffer);
+  } else {
+    DEBUG ((DEBUG_INFO, "OCC: Found NO handles with GOP draw\n"));
+  }
+
+  Status = gBS->LocateHandleBuffer (
+    ByProtocol,
+    &gEfiUgaDrawProtocolGuid,
+    NULL,
+    &HandleCount,
+    &HandleBuffer
+    );
+
+  if (!EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCC: Found %u handles with UGA for GOP check\n", (UINT32) HandleCount));
+
+    for (Index = 0; Index < HandleCount; ++Index) {
+      DEBUG ((DEBUG_INFO, "OCC: Trying handle %u - %p\n", (UINT32) Index, HandleBuffer[Index]));
+
+      Status = gBS->HandleProtocol (
+        HandleBuffer[Index],
+        &gEfiUgaDrawProtocolGuid,
+        (VOID **) &UgaDraw
+        );
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_INFO, "OCC: No UGA protocol - %r\n", Status));
+        continue;
+      }
+
+      Status = gBS->HandleProtocol (
+        HandleBuffer[Index],
+        &gEfiGraphicsOutputProtocolGuid,
+        (VOID **) &GraphicsOutput
+        );
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_INFO, "OCC: No GOP protocol - %r\n", Status));
+
+        FramebufferBase = 0;
+        FramebufferSize = 0;
+        ScreenRowBytes  = 0;
+        PixelFormat     = PixelBltOnly;
+
+        Status = gBS->HandleProtocol (
+          HandleBuffer[Index],
+          &gAppleFramebufferInfoProtocolGuid,
+          (VOID **) &FramebufferInfo
+          );
+        if (!EFI_ERROR (Status)) {
+          Status = FramebufferInfo->GetInfo (
+            FramebufferInfo,
+            &FramebufferBase,
+            &FramebufferSize,
+            &ScreenRowBytes,
+            &ScreenWidth,
+            &ScreenHeight,
+            &ScreenDepth
+            );
+          if (!EFI_ERROR (Status)) {
+            PixelFormat = PixelBlueGreenRedReserved8BitPerColor;
```

Comment:
Try RGBA maybe? I wonder if it is the reason it still fails.

Reply:
What is RGBA? Sorry, I do not see that [here](https://github.com/acidanthera/audk/blob/06dc822d045c2bb42e497487935485302486e151/MdePkg/Include/Protocol/GraphicsOutput.h#L28-L57).

Reply:
`PixelRedGreenBlueReserved8BitPerColor`, sorry

Reply:
Done.